update github status to allow for drier object oriented usage

jebentier opened the pull request:
Doing this to help with dryness in pipeline files that are looking to set many statuses for the same sha

jebentier committed:
update github status to allow for drier object oriented usage

jadametz reviewed (commented):
Can you provide a brief example of what calling this in a pipeline would look like?

jebentier commented:
Yes, example new usage is
```
githubStatusChecks = new GitHubStatus(this, 'Invoca/web', env.GIT_SHA,  env.BUILD_URL, env.TOKEN)
githubStatusChecks.setStatus('jenkins-ci/production-assets-pushed', 'Assets pushed to s3://invoca-assets', 'success')
```
or you can still use the old
```
gitHubStatus([
    repoSlug:    'Invoca/web',
    sha:         env.GIT_COMMIT,
    description: 'Assets pushed to s3://invoca-assets',
    context:     'jenkins-ci/production-assets-pushed',
    targetURL:   env.BUILD_URL,
    token:       env.GITHUB_TOKEN,
    status:      'success'
])
```

jebentier commented:
Yes, example new usage is
```
githubStatusChecks = new GitHubStatus(this, '<repo>', env.GIT_SHA,  env.BUILD_URL, env.TOKEN)
githubStatusChecks.setStatus('jenkins-ci/production-assets-pushed', 'Assets pushed to S3', 'success')
```
or you can still use the old
```
gitHubStatus([
    repoSlug:    '<repo>',
    sha:         env.GIT_COMMIT,
    description: 'Assets pushed to S3',
    context:     'jenkins-ci/production-assets-pushed',
    targetURL:   env.BUILD_URL,
    token:       env.GITHUB_TOKEN,
    status:      'success'
])
```

jebentier commented:
is there a contrib anywhere on how to apply contributions and make them public?

cmparkinson commented on src/com/invoca/github/GitHubStatus.groovy:
```diff
@@ -18,11 +18,15 @@ class GitHubStatus implements Serializable {
   }
 
-  void update() {
+  public void setStatus(String context, String status, String description) {
```
`GitHubStatus` objects are meant to encapsulate an individual status.  `context` is part of the composite key that identifies each status, so it should be included in the constructor.

Nit: `public` is the default in Groovy.

jebentier committed:
TECH-2610: extract out GitHubApi into its own class

jebentier commented:
@cmparkinson after an offline talk i went ahead and extracted the github api portion out into it's own object, which allows for things to be dried up.  I also switched the instance method `update` to take the new status as apposed to having the status be a part of the `GitHubStatus` object.  This feel like a better contract given our discussion about the idea of getting the current status from the API.

original execution of gitHubStatus variable remains working, and drier usage of GitHubStatus looks like:
```
github_status = new GitHubStatus([
    repoSlug:    '<repo>',
    sha:         env.GIT_COMMIT,
    description: 'Assets pushed to S3',
    context:     'jenkins-ci/production-assets-pushed',
    targetURL:   env.BUILD_URL,
    token:       env.GITHUB_TOKEN
])

github_status.update('success')
github_status.update('failure')
```

jebentier committed:
TECH-2610: remove new from string formatting

jebentier committed:
modify gitHubStatus to return the GitHubStatus object

jebentier committed:
call this instead of new

jebentier committed:
create fromConfig method for generating new GitHubStatus object from a config map

jebentier committed:
revert gitHubStatus back to its original functionality

jebentier commented:
@cmparkinson @jadametz am i good to merge this? and what's the release process?

jadametz commented:
Looks good to merge. No formally outlined contribution guidelines at the moment. To make it available it just needs to have a tag/release. Feel free to add release notes 😄
